src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/Bedrock/InvokeModelAsyncWrapper.cs

```diff
@@ -0,0 +1,340 @@
+using NewRelic.Reflection;
+using System.Net;
+using Newtonsoft.Json;
+
+namespace NewRelic.Providers.Wrapper.Bedrock
+{
+    public class InvokeModelAsyncWrapper : IWrapper
+    {
+        public bool IsTransactionRequired => true; // part of spec, only create events for transactions.
+
+        private static ConcurrentDictionary<Type, Func<object, object>> _getResultFromGenericTask = new();
+        private static ConcurrentDictionary<string, string> _libraryVersions  = new();
+        private const string WrapperName = "BedrockInvokeModelAsync";
+
+        public CanWrapResponse CanWrap(InstrumentedMethodInfo methodInfo)
+        {
+            return new CanWrapResponse(WrapperName.Equals(methodInfo.RequestedWrapperName));
+        }
+
+        public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall instrumentedMethodCall, IAgent agent, ITransaction transaction)
+        {
+            if (instrumentedMethodCall.IsAsync)
+            {
+                transaction.AttachToAsync();
+                transaction.DetachFromPrimary(); //Remove from thread-local type storage
+            }
+
+            dynamic invokeModelRequest = instrumentedMethodCall.MethodCall.MethodArguments[0];
+            var operationType = invokeModelRequest.ModelId.Contains("embed") ? "embedding" : "completion";
+            var segment = transaction.StartCustomSegment(
+                instrumentedMethodCall.MethodCall,
+                "Llm/" + operationType + "/Bedrock/" + instrumentedMethodCall.MethodCall.Method.MethodName
+            );
+
+            // required per spec
+            var version = GetOrAddLibraryVersion(instrumentedMethodCall.MethodCall.Method.Type.Assembly.ManifestModule.Assembly.FullName);
+            agent.RecordSupportabilityMetric("DotNet/ML/Bedrock/" + version);
+
+            return Delegates.GetAsyncDelegateFor<Task>(
+                agent,
+                segment,
+                false,
+                InvokeTryProcessResponse,
+                TaskContinuationOptions.ExecuteSynchronously
+            );
+
+            void InvokeTryProcessResponse(Task responseTask)
+            {
+                if (responseTask.IsFaulted)
+                {
+                    HandleError(segment, invokeModelRequest, responseTask, agent);
+                    segment.End();
+                    return;
+                }
+
+                // We need the duration, so we end the segment before creating the events.
+                segment.End();
+
+                dynamic invokeModelResponse = GetTaskResult(responseTask);
+                if (invokeModelResponse == null || invokeModelResponse.HttpStatusCode >= HttpStatusCode.MultipleChoices)
+                {
+                    agent.Logger.Log(Agent.Extensions.Logging.Level.Warn, $"Error invoking model {invokeModelRequest.ModelId}: Response payload {(invokeModelResponse == null ? "is null" : $"has non-success HttpStatusCode: {invokeModelResponse.HttpStatusCode}")}");
```

Comment:
That's a fair question. And I have no idea what the answer is... @jaffinito @chynesNR any ideas?

Reply:
I don't think it would be caught by our HTTP instrumentation, but as far as I can tell `InvokeModel` should always throw if there's an error, so it would be handled just above this. I don't think it's possible for it to return null, or not-throw but return an error status. The documentation is a little vague, though, so I'm not 100% sure.